### ISSUE TITLE
Exporting files?

### ISSUE THREAD

**Issue by HGustavs** [felixrieseberg/windows95#133]
A question.

It does not seem like the application saves to the image file or is there another image file that is used for saving?

I can install new applications and save new files but I have no way of accessing these files from outside windows95, as the mounted image only contains the files written to the image from outside the windows95 app.

**Comment by beedell-roke** [comment]
Yeah, I really want to know how to export files too, else this app is useless to me. I want to use it to extract old binaries from the Windows 95 image.

**Comment by HGustavs** [comment]
This app is so close to be genuinely useful for actual work.

I have thousands of documents that can only be opened from an app that could last be run on windows 95. I would be extremely happy if there was a way to export those files in some manner....

**Comment by BaRRaKudaRain** [comment]
Currently you can just open hard drive image file as archive and extract it. 

--
------------------------------
An outsider, the BaRRaKudaRain

-----Original Message-----

**Comment by HGustavs** [comment]
Which file is that?

The image file used by the system for booting does not seem to be written back into. If I mount the boot image file this image file remains unchanged compared to before booting. My question is where the running system keeps its file if there even is an image file for the running system?

**Comment by felixrieseberg** [comment]
Ah, accessing new files from within the system is pretty tough. It's true, we don't write directly to the disk image; instead, new stuff is saved in a state file.

You can find the source for the state stuff [here](https://github.com/copy/v86/blob/f6ae3ea58aee15f2a9be71f2847738367ef31747/src/state.js) and I'd love some help with this, but I don't think we'll have an easy way of getting to the files.

It might be easier to get networking to work, but I've had real trouble with that.

**Comment by pixeltris** [comment]
https://github.com/pixeltris/ww32vmt this may be of interest to you. It allows for partial internet access, transferring files to / from the VM, and clipboard access (text only). I haven't tested the internet access stuff on W95 yet but as long as you have TCP/IP installed (and possibly Winsock2) it should work.

The implementation is a little fragile and you may not be able to get it working after RAM starts filling up (if you start vmtools.exe early on and don't close it, it should be fine). Idealy this would be reimplemented using a COM port for communication (my goal was support for arbitrary web x86 VMs, which makes using COM ports a little more difficult without changing the internals of the emulator).

**Comment by blackstonespecs** [comment]
https://github.com/felixrieseberg/windows95/issues/144

If you're using WinXP or higher, your C:\ is most likely NTFS format, which can not be read at all by Windows 9x.
Check the link above for my quick fix. You CAN add files to the windows95.img file, but it has to be done properly or you risk corrupting it.

**Comment by cyco** [comment]
Hey, I had to run a 16-bit NE installer on my arm mac today and this project seemed like the easiest way to do that.

Since exporting files is not supported and I didn't manage to build a compatible v86 version to patch, I've decided to go the quick and dirty route and use electron dev tools to export the filesystem changes.
 
Maybe it's useful for somebody else as well:

```js
// download changes made to the file system, run this in dev tools
const download = (content,filename) => {
	const downloadLink = document.createElement("a");
	downloadLink.setAttribute("href", window.URL.createObjectURL(new Blob([content], { "application/octet-stream" })));
	downloadLink.setAttribute("download", filename);
	downloadLink.click();
};

download(JSON.stringify(emulator.disk_images.hda.get_written_blocks().indices), "indices.json");
download(emulator.disk_images.hda.get_written_blocks().buffer, "buffer");
```

Then to build a new image that can be mounted adjust the variables and run this in _node_:
```js
const FS = require("fs");

const blockSize = 0x100;
const inputPath = "path/to/boot/image/windows95.img";
const outputPath = "path/to/new/image.img";

const indicesPath = "path/to/downloaded/indices.json";
const changesPath = "path/to/downloaded/buffer";

const input = FS.readFileSync(inputPath);
const indices = JSON.parse(FS.readFileSync(indicesPath));
const changes = FS.readFileSync(changesPath);

for(let i=0; i < indices.length; i++) {
	for(let j=0; j < blockSize; j++) {
		input[indices[i] * blockSize + j] = changes[i * blockSize + j];
	}
}
```

**Comment by beedell-roke** [comment]
@ghost, I am not much of a developer, so I do not understand most of that. Consequently, does your code allow for creation of backups of the disk image?

**Comment by felixrieseberg** [comment]
Your read of the situation was right — and here's where things stand in the current release. Getting files *out* is more limited today:

- Text travels over the shared clipboard (copy inside Windows 95, paste on your host).
- Your changes live in a saved-state overlay, not in `windows95.img` — that's why the mounted image never shows them.
- The `Z:` share is currently read-only from the guest, by design.

Making `Z:` writable so you can copy files out is tracked in #321 — closing this one in favor of that.

**Comment by beedell-roke** [comment]
@felixrieseberg, probably better to close *as a duplicate*, then.
